pkg/apis/metal3/v1alpha1/baremetalhost_types.go

```diff
@@ -104,6 +109,49 @@ type BMCDetails struct {
 	CredentialsName string `json:"credentialsName"`
 }
 
+type RAIDVolume struct {
+	// Size (Integer) of the logical disk to be created in GiB.  If unspecified, "MAX" will be used.
+	SizeGB *int `json:"sizeGB"`
+
+	// RAID level for the logical disk.
+	RAIDLevel string `json:"raidLevel" required:"true"`
+
+	// Name of the volume. Should be unique within the Node. If not specified, volume name will be auto-generated.
+	VolumeName string `json:"volumeName,omitempty"`
+
+	// Set to true if this logical disk can share physical disks with other logical disks.
+	SharePhysicalDisks *bool `json:"sharePhysicalDisks,omitempty"`
+
+	// If this is not specified, disk type will not be a criterion to find backing physical disks
+	DiskType string `json:"diskType,omitempty"`
+
+	// If this is not specified, interface type will not be a criterion to find backing physical disks.
+	InterfaceType string `json:"interfaceType,omitempty"`
+
+	// Integer, number of disks to use for the logical disk. Defaults to minimum number of disks required
+	// for the particular RAID level.
+	NumberOfPhysicalDisks int `json:"numberOfPhysicalDisks,omitempty"`
+
+	// The name of the controller as read by the RAID interface.
+	Controller string `json:"controller,omitempty"`
+
+	// A list of physical disks to use as read by the RAID interface.
+	PhysicalDisks []string `json:"physicalDisks,omitempty"`
+}
+
+// RAIDConfig contains the configuration that are required to config RAID in Bare Metal server
+type RAIDConfig struct {
+
+	// The logical disk that will be root volume
+	RootVolume *RAIDVolume `json:"rootVolume,omitempty"`
+
+	// The list of logical disks
+	Volumes []RAIDVolume `json:"volumes,omitempty"`
+}
+
+// BIOSConfig contains the configuration that are required to config BIOS in Bare Metal server
+type BIOSConfig map[string]interface{}
```

Comment:
Because BIOS setting depend on vendor's driver, so I declare `BIOSConfig` as `interface{}` to fit with all of BIOS struct from vendors, then each vendor have to defined `VendorBIOSConfigSpec` [1] for their own BIOS settings.

For BIOS validation, I also defined `ValueType` for each BIOS setting in `VendorBIOSConfigSpec`, so we can validate the BIOS setting from users via `ValidateAndBuildBIOSCleanStep` function.

[1] https://github.com/metal3-io/baremetal-operator/pull/279/files#diff-334e60af5353ff0c1ef95211da871902R53
[2] https://github.com/metal3-io/baremetal-operator/pull/279/files#diff-bc78df50e156a9dedd77791eea452d13R1085

Reply:
I realize it is going to make it more complicated to design an API that supports changing different settings on different types of hardware, but that's the task. A pass-through API is completely untenable, because it's *not* an API.

We *must* decide on some way to describe every API parameter. That may mean abstracting some of the BIOS settings so the same inputs can apply on different types of hardware. It may mean providing a way to specify hardware-type-specific BIOS settings. We will not have an API that accepts an unspecified input and relies on validation happening after the fact.